### PR TITLE
Unique minimum of totally-ordered sets

### PR DESCRIPTION
**Pull request by stefan-aws**
Establishes that every totally-ordered set has a unique minimum.

The proof is a simplified version of the arguments in https://leino.science/papers/krml275.html.

By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

**Review by MikaelMayer — DISMISSED** [review]
Straightforward, good job !

File: src/Relations.dfy [hunk 1]
```diff
@@ -49,6 +49,27 @@ module {:options "-functionSyntax:4"} Relations {
...
 
+  /* An element in an ordered set is called minimal, if it is less than every element of the set. */
+  ghost predicate IsMinimum<T>(R: (T, T) -> bool, m: T, s: set<T>) {
+    m in s && forall y: T | y in s :: R(m, y)
+  }
+
+  /* Any totally ordered set contains a unique minimal element. */
+  lemma LemmaUniqueMinimum<T(!new)>(R: (T, T) -> bool, s: set<T>) returns (m: T)
+    requires |s| > 0 && TotalOrdering(R)
+    ensures IsMinimum(R, m, s) && (forall n: T :: IsMinimum(R, n, s) ==> m == n)
```
Comment (jtristan) [inline src/Relations.dfy:60]: I think that Rustan prefers that we use the syntactic sugar: 
```
forall n: T | IsMinimum(R, n, s) :: m == n
```

File: src/Relations.dfy [hunk 2]
```diff
@@ -49,6 +49,27 @@ module {:options "-functionSyntax:4"} Relations {
...
+  {
+    var x :| x in s;
```
Comment (jtristan) [inline src/Relations.dfy:62]: I wonder if we should try to type variables explicitly? Is there some guideline on whether or not we want this for the library?

File: src/Relations.dfy [hunk 3]
```diff
@@ -49,6 +49,27 @@ module {:options "-functionSyntax:4"} Relations {
...
+    if s == {x} {
+      m := x;
+    } else {
+      var m' := LemmaUniqueMinimum(R, s - {x});
+      if
```
Comment (jtristan) [inline src/Relations.dfy:67]: I don't really like that syntax so much, but it's just a matter of taste

**Review by jtristan — DISMISSED** [review]

File: src/Relations.dfy (hunk 2 above)
Comment (stefan-aws) [reply src/Relations.dfy:62]: The style guide http://dafny.org/dafny/StyleGuide/Style-Guide.html#type-declaration contains the guidance "If the type can be inferred by Dafny, leave it out, unless you think it provides useful documentation in the program.".

**Review by jtristan — APPROVED** [review]

**Review by MikaelMayer — APPROVED** [review]

**Pull request merged**